Source/Core/DolphinQt/FIFO/FIFOAnalyzer.cpp

```diff
@@ -164,24 +173,124 @@ void FIFOAnalyzer::UpdateTree()
-  static constexpr std::array<const char*, 8> names = {
-      "GX_DRAW_QUADS",        "GX_DRAW_QUADS_2 (nonstandard)",
-      "GX_DRAW_TRIANGLES",    "GX_DRAW_TRIANGLE_STRIP",
-      "GX_DRAW_TRIANGLE_FAN", "GX_DRAW_LINES",
-      "GX_DRAW_LINE_STRIP",   "GX_DRAW_POINTS",
-  };
-  return fmt::format("{} VAT {}", names[primitive], vat);
-}
+
+  void OnXF(u16 address, u8 count, const u8* data) override
+  {
+    const auto [name, desc] = GetXFTransferInfo(address, count, data);
+    ASSERT(!name.empty());
+
+    const u32 command = address | (count << 16);
+
+    text = QStringLiteral("XF  %1  ").arg(command, 8, 16, QLatin1Char('0'));
+
+    for (u8 i = 0; i < count; i++)
+    {
+      const u32 value = Common::swap32(&data[i * 4]);
+
+      text += QStringLiteral("%1 ").arg(value, 8, 16, QLatin1Char('0'));
+    }
+
+    text += QStringLiteral("  ") + QString::fromStdString(name);
+  }
+
+  void OnBP(u8 command, u32 value) override
+  {
+    const auto [name, desc] = GetBPRegInfo(command, value);
+    ASSERT(!name.empty());
+
+    text = QStringLiteral("BP  %1  %2  %3")
+               .arg(command, 2, 16, QLatin1Char('0'))
+               .arg(value, 6, 16, QLatin1Char('0'))
+               .arg(QString::fromStdString(name));
+  }
+  void OnIndexedLoad(u8 array, u32 index, u16 address, u8 size) override
+  {
+    const auto [desc, written] = GetXFIndexedLoadInfo(array, index, address, size);
+    text = QStringLiteral("LOAD INDX %1   %2")
+               .arg("ABCD"[array - ARRAY_XF_A])
+               .arg(QString::fromStdString(desc));
+  }
+  void OnPrimitiveCommand(OpcodeDecoder::Primitive primitive, u8 vat, u32 vertex_size,
+                          u16 num_vertices, const u8* vertex_data) override
+  {
+    const auto name = fmt::to_string(primitive);
+
+    // Note that vertex_count is allowed to be 0, with no special treatment
+    // (another command just comes right after the current command, with no vertices in between)
+    const u32 object_prim_size = num_vertices * vertex_size;
+
+    const u8 opcode = 0x80 | (u8(primitive) << OpcodeDecoder::GX_PRIMITIVE_SHIFT) | vat;
+    text = QStringLiteral("PRIMITIVE %1 (%2)  %3 vertices %4 bytes/vertex %5 total bytes")
+               .arg(QString::fromStdString(name))
+               .arg(opcode, 2, 16, QLatin1Char('0'))
+               .arg(num_vertices)
+               .arg(vertex_size)
+               .arg(object_prim_size);
+
+    // It's not really useful to have a massive unreadable hex string for the object primitives.
+    // Put it in the description instead.
+
+// #define INCLUDE_HEX_IN_PRIMITIVES
+#ifdef INCLUDE_HEX_IN_PRIMITIVES
+    text += QStringLiteral("   ");
+    for (u32 i = 0; i < object_prim_size; i++)
+    {
+      text += QStringLiteral("%1").arg(vertex_data[i], 2, 16, QLatin1Char('0'));
+    }
+#endif
+  }
+
+  void OnDisplayList(u32 address, u32 size) override
+  {
+    // Should have been inlined by the recorder
+    ASSERT(false);
```

Comment:
Note that this can cause false-positives with code relating to static analysis, since this will always be executed and consider the following code as dead.

Reply:
Does that apply do Dolphin's `ASSERT`?  I'm pretty sure that Dolphin's one will return normally if the user chooses to ignore it.  (I can definitely see it causing problems with C's `abort` which is called by `assert` when it's enabled, since `abort` does not return)